### DISCUSSION TITLE
Mount option "dataCenter" (automatic?)

### DISCUSSION THREAD

**Discussion by Ruakij** [seaweedfs/seaweedfs-csi-driver#99] [Ideas]
I'd like to add the mount-option "dataCenter" from [weed mount](https://github.com/seaweedfs/seaweedfs/blob/master/weed/command/mount.go#L22).
It should be able to get added to _StorageClass_ parameters and _PersistentVolume_.

Is this just a matter of adding it to [the mounter](https://github.com/seaweedfs/seaweedfs-csi-driver/blob/master/pkg/driver/mounter_seaweedfs.go#L63)?

And if so, where?
[This comment](https://github.com/seaweedfs/seaweedfs-csi-driver/blob/master/pkg/driver/mounter_seaweedfs.go#L84) gave me some hesitation, but i didnt find a better suited location while looking.
Is `volumeAttributes` from _PersistentVolume_ and `parameters` from _StorageClass_  both `volContext`?

---

I'd also like to add an automatic selector for the dataCenter based on the dataCenter the current pod is running in.
Node-Label `topology.kubernetes.io/zone` ([Documentation](https://kubernetes.io/docs/reference/labels-annotations-taints/#topologykubernetesiozone )) could be used?
Maybe have this option enabled by-default?

Is it ok to set this in [mounter_seaweedfs.go](https://github.com/seaweedfs/seaweedfs-csi-driver/blob/master/pkg/driver/mounter_seaweedfs.go)?

That would also pave the first step towards "Topology-aware volume provisioners".

Read:
Blog-Post:https://kubernetes.io/blog/2018/10/11/topology-aware-volume-provisioning-in-kubernetes/
K8s-Documentation: https://kubernetes.io/docs/concepts/storage/storage-classes/#volume-binding-mode
CSI-documentation: https://kubernetes-csi.github.io/docs/topology.html

**Comment by chrislusf** [comment]
> Is `volumeAttributes` from _PersistentVolume_ and `parameters` from _StorageClass_ both `volContext`?

Need a k8s expert on this.

**Reply by Ruakij** [reply]
Have tested, yes its the same.
Not sure why they named it differently..

**Reply by chrislusf** [reply]
seems the same thing but controlled at different level. 

storage class sounds a type of storage, while PersistentVolume seems a specific instance of the storage type.

**Comment by Ruakij** [comment]
I think there is a bug in mapping the volContext-parameters to mount-args?

In [mounter_seaweedfs.go](https://github.com/seaweedfs/seaweedfs-csi-driver/blob/master/pkg/driver/mounter_seaweedfs.go#L96), `diskType` is translated to `disk`, but the [mount-option for weed mount](https://github.com/seaweedfs/seaweedfs/blob/master/weed/command/mount.go#L53) is called `diskType`.

---

Is it okay if i remove the selective-mapping from Line 89 - 96 and simply add any `volContext` as `args`?
Special operations like for `volumeCapacity` will of course be kept.

This enables setting any mount-options (like dataCenter, currently missing) using the context without having to add them manually in the code.
Of course this also allows misconfiguration, but which can be found quickly using the container logs from seaweedfs-csi-driver-node.

**Reply by chrislusf** [reply]
you can remove 89-94 and add a default section.
